zebra: dataplane batching

xThaid opened the pull request:
Currently, in the dplane pthread we dequeue a single context object and send it to the kernel via suitable interface. This PR extends the current kernel API by the function that simply takes a list of context objects. Supplementary to this, I implemented the infrastructure for the batch sending messages to the linux kernel via netlink.

[This diagram](https://github.com/xThaid/frr/wiki/Dataplane-batching#code-flowchart) may help to understand the code flow.

I've also done [performance tests](https://github.com/xThaid/frr/wiki/Dataplane-batching#performance-tests).

~~This is rather draft than a final, ready to merge PR. I plan to add dev docs and some unit tests later.~~
It is a final PR.

mjstapp commented on zebra/kernel_netlink.c:
```diff
@@ -1080,6 +1132,288 @@ int netlink_request(struct nlsock *nl, void *req)
+	case DPLANE_OP_ROUTE_INSTALL:
+	case DPLANE_OP_ROUTE_UPDATE:
+	case DPLANE_OP_ROUTE_DELETE:
+		return netlink_put_route_update_msg(bth, ctx);
```
These were a little convoluted  - we add new apis that get called here, and then end up ... 50 lines above, with different encoder callbacks. why not just have the callbacks available, and skip the 'put_xxx_msg' layer here - is that providing some value?

xThaid replied:
Again, there is the problem with context objects that require two Netlink messages. I know, this looks confusing, but I couldn't came up with more sensible solution.

mjstapp replied:
ok, yes, I get what you're saying

xThaid commented on zebra/if_netlink.h:
```diff
@@ -32,6 +32,10 @@ extern int netlink_interface_addr(struct nlmsghdr *h, ns_id_t ns_id,
 extern int netlink_link_change(struct nlmsghdr *h, ns_id_t ns_id, int startup);
 extern int interface_lookup_netlink(struct zebra_ns *zns);
 
+struct nl_batch;
```
You're right, it isn't. I removed this and some other.

xThaid commented on zebra/zebra_dplane.c:
```diff
@@ -3847,25 +3802,51 @@ static int kernel_dplane_process_func(struct zebra_dplane_provider *prov)
 		TAILQ_INSERT_TAIL(&work_list, ctx, zd_q_entries);
 	}
 
+	if (kernel_supports_batch())
```
there's only one api now, so it should be clear.

xThaid commented:
I rebased onto master and removed 'one-at-a-time' api, so there is only one way to send the kernel update - by handling a list of context objects.

xThaid commented on zebra/rt_netlink.c:
```diff
@@ -2227,33 +2219,51 @@ enum zebra_dplane_result kernel_nexthop_update(struct zebra_dplane_ctx *ctx)
+ */
+enum zebra_dplane_result kernel_nexthop_update(struct zebra_dplane_ctx *ctx)
+{
+	return ZEBRA_DPLANE_REQUEST_FAILURE;
```
I agree. I just deleted the old api.

xThaid commented on zebra/if_netlink.c:
```diff
@@ -921,64 +922,80 @@ static int netlink_address_ctx(const struct zebra_dplane_ctx *ctx)
+ * The communication with the kernel is done using the message batching
+ * interface, so return a failure.
+ */
 enum zebra_dplane_result kernel_address_update_ctx(struct zebra_dplane_ctx *ctx)
```
Having removed the old api, this comment may be outdated.

xThaid commented on zebra/kernel_netlink.c:
```diff
@@ -1080,6 +1132,288 @@ int netlink_request(struct nlsock *nl, void *req)
+		 * time iterate over the batch item list and match responses
+		 * with requests.
+		 */
+		frr_each_from(nl_batch_list, &(bth->items), item, from_item) {
```
According to the docs the `frr_each_from` can resume iteration after breaking out of the loop by keeping the from value persistent and reusing it for the next loop. And that's what I wanted to do here. Since both `nl_batch_list` and Netlink response contains messages sorted by their sequence numbers, so we can match items using two pointers technique.

mjstapp replied:
ah, sure, sorry, I didn't read the 'from' macro well enough!

xThaid commented on zebra/kernel_netlink.c:
```diff
@@ -1080,6 +1132,288 @@ int netlink_request(struct nlsock *nl, void *req)
+}
+
+static void nl_batch_add_item(struct nl_batch *bth, int seq,
+			      struct zebra_dplane_ctx *ctx, bool ignore_res)
```
The problem is that for one context object there could be 0, 1 or 2 Netlink messages, so this `item` struct corresponds to every message that we sent to the kernel.

mjstapp replied:
hmm - but if there are 0 netlink messages, then  ... there's nothing to match? and in the case where there are two messages - the delete + re-add case, I guess? - it's only the second message that we care about, isn't it? I mean, you've already issued the two messages, you can't really react to any failure from the first message, so whatever happened... happened?

xThaid commented on zebra/kernel_netlink.c:
```diff
@@ -1080,6 +1132,288 @@ int netlink_request(struct nlsock *nl, void *req)
+		nl_batch_send(bth);
+
+	size = (*msg_encoder)(ctx, bth->buf_head, bth->bufsiz - bth->curlen);
+	if (size < 0)
```
I added enum as you proposed, so it should be more self-descriptive now. 
`encoder` returns zero for a special kind of failure - when the message wouldn't fit in the buffer. We need to distinguish this error from others, because we can recover from this error -  simply encode this message again to the empty buffer.

xThaid commented:
I added a test in which I resize the size of the buffer to a small number and then install 100 routes. For that, I also added a hidden CLI command that allows to change the size of the buffer used in the batching at runtime.

xThaid commented:
Rebased onto master and added docs.
